RecoJets/JetProducers/src/MVAJetPuId.cc

```diff
@@ -0,0 +1,568 @@
+		float candPtDr = candPt * candDr;
+		size_t icone = std::lower_bound(&cones[0],&cones[ncones],candDr) - &cones[0];
+		float weight2 = candPt * candPt;
+
+
+		if( lLead.isNull() || candPt > lLead->pt() ) {
+			lSecond = lLead;
+			lLead = icand; 
+		} else if( (lSecond.isNull() || candPt > lSecond->pt()) && (candPt < lLead->pt()) ) {
+			lSecond = icand;
+		}
+
+		//internalId_.dRMean_     += candPtDr;
+		internalId_.DR_weighted_    += candPtDr*candPtDr;
+
+		internalId_.ptD_ += candPt*candPt;
+		sumPt += candPt;
+		sumPt2 += candPt*candPt;
+		sum_deta     += candDeta*weight2;
+		sum_dphi     += candDphi*weight2;
+		sum_deta2    += candDeta*candDeta*weight2;
+		sum_detadphi += candDeta*candDphi*weight2;
+		sum_dphi2    += candDphi*candDphi*weight2;
+		//Teta         += candPt * candDR * candDeta;
+		//Tphi         += candPt * candDR * candDphi;
+
+
+
+		frac.push_back(candPtFrac);
+		if( icone < ncones ) { *coneFracs[icone] += candPt; }
+
+
+		if( icand->particleId() == reco::PFCandidate::h0 ) {
+			if (lLeadNeut.isNull() || candPt > lLeadNeut->pt()) { lLeadNeut = icand; }
+			internalId_.dRMeanNeut_ += candPtDr;
+			fracNeut.push_back(candPtFrac);
+			sumPtNe               += candPt;
+		}
+
+		if( icand->particleId() == reco::PFCandidate::gamma ) {
+			if(lLeadEm.isNull() || candPt > lLeadEm->pt())  { lLeadEm = icand; }
+			internalId_.dRMeanEm_ += candPtDr;
+			fracEm.push_back(candPtFrac);
+			sumPtNe               += candPt;
+		}
+
+		if(  icand->trackRef().isNonnull() && icand->trackRef().isAvailable() ) {
+			if (lLeadCh.isNull() || candPt > lLeadCh->pt()) { lLeadCh = icand; }
+			//internalId_.jetRchg_  += candPtDr;
+			fracCh.push_back(candPtFrac);
+			sumPtCh                += candPt;
+		}
+
+		if(  icand->trackRef().isNonnull() && icand->trackRef().isAvailable() ) {
+			try { 
+				float tkpt = icand->trackRef()->pt(); 
+				sumTkPt += tkpt;
+				bool inVtx0 = find( vtx->tracks_begin(), vtx->tracks_end(), reco::TrackBaseRef(icand->trackRef()) ) != vtx->tracks_end();
+				bool inAnyOther = false;
+
+				double dZ0 = fabs(icand->trackRef()->dz(vtx->position()));
+				double dZ = dZ0;
+				for(reco::VertexCollection::const_iterator  vi=allvtx.begin(); vi!=allvtx.end(); ++vi ) {
+					const reco::Vertex & iv = *vi;
+					if( iv.isFake() || iv.ndof() < 4 ) { continue; }
+
+					bool isVtx0  = (iv.position() - vtx->position()).r() < 0.02;
+
+					if( ! isVtx0 && ! inAnyOther ) {
+						inAnyOther = find( iv.tracks_begin(), iv.tracks_end(), reco::TrackBaseRef(icand->trackRef()) ) != iv.tracks_end();
+					}
+
+					dZ = std::min(dZ,fabs(icand->trackRef()->dz(iv.position())));
+				}
+				if( inVtx0 && ! inAnyOther ) {
+					internalId_.betaClassic_ += tkpt;
+				} else if( ! inVtx0 && inAnyOther ) {
+					internalId_.betaStarClassic_ += tkpt;
+				}
+
+				if( dZ0 < 0.2 ) {
+					internalId_.beta_ += tkpt;
+				} else if( dZ < 0.2 ) {
+					internalId_.betaStar_ += tkpt;
+				}
+			} catch (cms::Exception &e) {
```

Comment:
all used code functions are under control here: try-catch should be removed and issue should be handled in a different way if it exists.